server.h

```diff
@@ -278,8 +297,50 @@ class CTFOServer {
   }
 
   void OnMidichloriansEvent(const LogEntry& entry) {
-    // TODO(mzhurovich): update answers here.
-    static_cast<void>(entry);
+    std::unique_ptr<MidichloriansEvent> event;
+    if (entry.m == "POST") {
+      try {
+        ParseJSON(entry.b, event);
+        UpdateStateOnEvent(event);
+      } catch (const bricks::ParseJSONException&) {
+        // TODO(mzhurovich): Error logging.
+        std::cerr << "Parse JSON failed." << std::endl;
+      }
+    }
+  }
+
+  void UpdateStateOnEvent(const std::unique_ptr<MidichloriansEvent>& event) {
+    try {
+      const iOSGenericEvent& ge = dynamic_cast<const iOSGenericEvent&>(*event.get());
+      if (valid_answers_.find(ge.event) != valid_answers_.end()) {
+        const ANSWER answer = valid_answers_.at(ge.event);
+        const UID uid = StringToUID(ge.fields.at("uid"));
+        const CID cid = StringToCID(ge.fields.at("cid"));
+        const std::string token = ge.fields.at("token");
+        if (uid != UID::INVALID && cid != CID::INVALID) {
+          storage_.Transaction([&uid, &cid, &token, &answer](StorageAPI::T_DATA data) {
+                                 const auto auth_token_accessor = Matrix<AuthKeyTokenPair>::Accessor(data);
+                                 bool token_is_valid = false;
+                                 if (auth_token_accessor.Cols().Has(token)) {
+                                   // Something went terribly wrong
+                                   // if we have more than one authentication key for token.
+                                   assert(auth_token_accessor[token].size() == 1);
+                                   if (auth_token_accessor[token].begin()->valid) {
+                                     token_is_valid = true;
+                                   }
+                                 }
+                                 if (token_is_valid) {
+                                   auto answers_mutator = Matrix<Answer>::Mutator(data);
+                                   if (!answers_mutator.Has(uid, cid)) {  // Do not overwrite existing answers.
+                                     data.Add(Answer(uid, cid, answer));
+                                   }
+                                 }
+                               }).Go();
+        }
+      }
+    } catch (const std::bad_cast&) {
```

Comment:
And, by all means, keep the `catch` for `std::bad_cast`. Sorry and my bad once again.